File: Algorithms/p331_Verify_Preorder_Serialization_of_a_Binary_Tree.py
class Solution(object):
    def isValidSerialization(self, preorder):
        """
        :type preorder: str
        :rtype: bool
        """
        if not preorder:
            return False
        cnt = 1
        pos = 0
        while pos!=-1 and cnt>0:
            pos = preorder.find(',',pos+1)
            if (pos!=-1 and preorder[pos-1]=='#') or (pos==-1 and preorder[-1]=='#'):
                cnt -= 1
            else:
                cnt += 1
        return pos==-1 and cnt==0

File: Algorithms/p331_Verify_Preorder_Serialization_of_a_Binary_Tree/p331_Verify_Preorder_Serialization_of_a_Binary_Tree.cpp
class Solution {
public:
    bool isValidSerialization(string preorder) {
        if(preorder.empty()) return 0;
        int cnt = 1;
        int pos = 0;
        while(pos!=string::npos&&cnt>0){
            pos = preorder.find_first_of(',',pos+1);
            if(pos!=string::npos&&preorder[pos-1]=='#' || pos==string::npos&&preorder[preorder.size()-1]=='#'){
                --cnt;
            }else{
                ++cnt;
            }
        }
        return pos==string::npos&&cnt==0;
    }
};
